In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
import sys
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/enron-email-dataset/emails.csv
/kaggle/input/final-project-datasetpkl/final_project_dataset.pkl


Loading data

In [2]:
import pickle
original = "/kaggle/input/final-project-datasetpkl/final_project_dataset.pkl"
destination = "modified_final_project_dataset.pkl"

content = ''
outsize = 0
with open(original, 'rb') as infile:
    content = infile.read()
with open(destination, 'wb') as output:
    for line in content.splitlines():
        outsize += len(line) + 1
        output.write(line + str.encode('\n'))

In [3]:
data_dict = pickle.load(open("modified_final_project_dataset.pkl", 'rb') )

In [4]:
all_features = []
temp = 0
for key in data_dict:
    if temp == 0:
        for feature in data_dict[key]:
            all_features.append(feature)
        temp = 1
print ("Features:" + "\n"+ str(all_features))
print ("Number of features:" +str(len(all_features)))

Features:
['salary', 'to_messages', 'deferral_payments', 'total_payments', 'loan_advances', 'bonus', 'email_address', 'restricted_stock_deferred', 'deferred_income', 'total_stock_value', 'expenses', 'from_poi_to_this_person', 'exercised_stock_options', 'from_messages', 'other', 'from_this_person_to_poi', 'poi', 'long_term_incentive', 'shared_receipt_with_poi', 'restricted_stock', 'director_fees']
Number of features:21


**removing some data**

In [5]:
data_dict.pop('THE TRAVEL AGENCY IN THE PARK')



{'salary': 'NaN',
 'to_messages': 'NaN',
 'deferral_payments': 'NaN',
 'total_payments': 362096,
 'loan_advances': 'NaN',
 'bonus': 'NaN',
 'email_address': 'NaN',
 'restricted_stock_deferred': 'NaN',
 'deferred_income': 'NaN',
 'total_stock_value': 'NaN',
 'expenses': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'exercised_stock_options': 'NaN',
 'from_messages': 'NaN',
 'other': 362096,
 'from_this_person_to_poi': 'NaN',
 'poi': False,
 'long_term_incentive': 'NaN',
 'shared_receipt_with_poi': 'NaN',
 'restricted_stock': 'NaN',
 'director_fees': 'NaN'}

In [6]:
data_dict.pop('LOCKHART EUGENE E')

{'salary': 'NaN',
 'to_messages': 'NaN',
 'deferral_payments': 'NaN',
 'total_payments': 'NaN',
 'loan_advances': 'NaN',
 'bonus': 'NaN',
 'email_address': 'NaN',
 'restricted_stock_deferred': 'NaN',
 'deferred_income': 'NaN',
 'total_stock_value': 'NaN',
 'expenses': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'exercised_stock_options': 'NaN',
 'from_messages': 'NaN',
 'other': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'poi': False,
 'long_term_incentive': 'NaN',
 'shared_receipt_with_poi': 'NaN',
 'restricted_stock': 'NaN',
 'director_fees': 'NaN'}

In [7]:
data_dict.pop("TOTAL")

{'salary': 26704229,
 'to_messages': 'NaN',
 'deferral_payments': 32083396,
 'total_payments': 309886585,
 'loan_advances': 83925000,
 'bonus': 97343619,
 'email_address': 'NaN',
 'restricted_stock_deferred': -7576788,
 'deferred_income': -27992891,
 'total_stock_value': 434509511,
 'expenses': 5235198,
 'from_poi_to_this_person': 'NaN',
 'exercised_stock_options': 311764000,
 'from_messages': 'NaN',
 'other': 42667589,
 'from_this_person_to_poi': 'NaN',
 'poi': False,
 'long_term_incentive': 48521928,
 'shared_receipt_with_poi': 'NaN',
 'restricted_stock': 130322299,
 'director_fees': 1398517}

In [8]:
def make_Fraction(_messages, poi_messages):

    if _messages != 'NaN'or poi_messages != 'NaN':
        fraction = float(poi_messages) / float(_messages)
    else:
        fraction = 0
    return fraction

In [9]:
 for name in data_dict:
    data = data_dict[name]
    
    from_poi_to_this_person = data["from_poi_to_this_person"]
    to_messages = data["to_messages"]
    fraction_from_poi = make_Fraction( to_messages, from_poi_to_this_person)
    
    data["fraction_from_poi"] = fraction_from_poi
    
    from_this_person_to_poi = data["from_this_person_to_poi"]
    from_messages = data["from_messages"]
    fraction_to_poi = make_Fraction( from_messages, from_this_person_to_poi)
    
    data["fraction_to_poi"] = fraction_to_poi

In [10]:
new_features = []
temp = 0
for key in data_dict:
    if temp == 0:
        for feature in data_dict[key]:
            new_features.append(feature)
        temp = 1

In [11]:
len(new_features)

23

In [12]:
new_features

['salary',
 'to_messages',
 'deferral_payments',
 'total_payments',
 'loan_advances',
 'bonus',
 'email_address',
 'restricted_stock_deferred',
 'deferred_income',
 'total_stock_value',
 'expenses',
 'from_poi_to_this_person',
 'exercised_stock_options',
 'from_messages',
 'other',
 'from_this_person_to_poi',
 'poi',
 'long_term_incentive',
 'shared_receipt_with_poi',
 'restricted_stock',
 'director_fees',
 'fraction_from_poi',
 'fraction_to_poi']

In [13]:
features_remove = ["poi", "email_address", "from_messages", "to_messages", "from_poi_to_this_person", "from_this_person_to_poi"]

In [14]:
modified_feature = ["poi"]
for feature in new_features:
    if feature not in features_remove:
       modified_feature.append(feature)
print ("features_list = {}".format(modified_feature))

features_list = ['poi', 'salary', 'deferral_payments', 'total_payments', 'loan_advances', 'bonus', 'restricted_stock_deferred', 'deferred_income', 'total_stock_value', 'expenses', 'exercised_stock_options', 'other', 'long_term_incentive', 'shared_receipt_with_poi', 'restricted_stock', 'director_fees', 'fraction_from_poi', 'fraction_to_poi']


In [15]:
len(modified_feature)

18

Splitting Data

In [16]:

import numpy as np

def featureFormat( dictionary, features, remove_NaN=True, remove_all_zeroes=True, remove_any_zeroes=False, sort_keys = False):
    
    return_list = []

    # Key order - first branch is for Python 3 compatibility on mini-projects,
    # second branch is for compatibility on final project.
    if isinstance(sort_keys, str):
        import pickle
        keys = pickle.load(open(sort_keys, "rb"))
    elif sort_keys:
        keys = sorted(dictionary.keys())
    else:
        keys = dictionary.keys()

    for key in keys:
        tmp_list = []
        for feature in features:
            try:
                dictionary[key][feature]
            except KeyError:
                print ("error: key ", feature, " not present")
                return
            value = dictionary[key][feature]
            if value=="NaN" and remove_NaN:
                value = 0
            tmp_list.append( float(value) )

        # Logic for deciding whether or not to add the data point.
        append = True
        # exclude 'poi' class as criteria.
        if features[0] == 'poi':
            test_list = tmp_list[1:]
        else:
            test_list = tmp_list
        ### if all features are zero and you want to remove
        ### data points that are all zero, do that here
        if remove_all_zeroes:
            append = False
            for item in test_list:
                if item != 0 and item != "NaN":
                    append = True
                    break
        ### if any features for a given data point are zero
        ### and you want to remove data points with any zeroes,
        ### handle that here
        if remove_any_zeroes:
            if 0 in test_list or "NaN" in test_list:
                append = False
        ### Append the data point if flagged for addition.
        if append:
            return_list.append( np.array(tmp_list) )

    return np.array(return_list)



def targetFeatureSplit( data ):
    """ 
        given a numpy array like the one returned from
        featureFormat, separate out the first feature
        and put it into its own list (this should be the 
        quantity you want to predict)

        return targets and features as separate lists

        (sklearn can generally handle both lists and numpy arrays as 
        input formats when training/predicting)
    """

    target = []
    features = []
    for item in data:
        target.append( item[0] )
        features.append( item[1:] )

    return target, features

In [17]:
dataset = data_dict
data = featureFormat(dataset, modified_feature, sort_keys = True)
Y, X= targetFeatureSplit(data)

In [18]:
import numpy as np
from sklearn.model_selection import StratifiedShuffleSplit

In [19]:
X = np.array(X)
Y = np.array(Y)
cv = StratifiedShuffleSplit(n_splits=1000, random_state=42)
for train, test in cv.split(X, Y):
    X_train, X_test = X[train], X[test]
    Y_train, Y_test = Y[train], Y[test]

In [20]:
print(len(X_train),len(X_test),len(Y_train), len(Y_test))

128 15 128 15


preprocessing

In [21]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.feature_selection import SelectKBest, chi2

In [22]:
min_max_scaler = MinMaxScaler()
X_train_minmax = min_max_scaler.fit_transform(X_train)
X_test_minmax = min_max_scaler.transform(X_test)

In [23]:
select = SelectKBest(chi2, k=5) 
select.fit(X_train_minmax, Y_train)
X_train_minmax_skb = select.transform(X_train_minmax)
X_test_minmax_skb = select.transform(X_test_minmax)

In [24]:
select.scores_

array([1.79853834, 0.07301844, 0.52588964, 0.17142857, 2.62543546,
       0.00768846, 0.15128769, 3.90756737, 1.35844284, 3.62891357,
       0.00516472, 0.63930146, 1.58438395, 1.56354953, 1.36026087,
       0.6283748 , 2.3993314 ])

## Selecting best model

***Importing modules***

In [25]:
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, recall_score, precision_score
from sklearn.metrics import accuracy_score

explore various classifiers

In [26]:
names = ["K Nearest Neighbors", "RBF SVM", "Linear SVM", "Decision Tree", "Naive Bayes", 
         "AdaBoost", "Random Forest", "Logistic Regression"]

In [27]:
classifiers = [KNeighborsClassifier(), SVC(kernel='rbf', random_state=42), \
                SVC(kernel="linear", random_state=42), DecisionTreeClassifier(random_state=42), \
                GaussianNB(), AdaBoostClassifier(random_state=42), RandomForestClassifier(random_state=42), \
                LogisticRegression(random_state=42)]

In [28]:
confusion_matrices = []
precisionscores = []
recallscores = []
accuracy = []
table={}
for name, clf in zip(names, classifiers):
    clf.fit(X_train_minmax_skb, Y_train)
    pred = clf.predict(X_test_minmax_skb)
    conf_mat = confusion_matrix(Y_test, pred)
    confusion_matrices.append(conf_mat)
    precisionscore = round(precision_score(Y_test, pred), 3)
    precisionscores.append(precisionscore)
    recallscore = round(recall_score(Y_test, pred), 3)
    recallscores.append(recallscore)
    accuracy.append(accuracy_score(Y_test, pred))
    
    _scores = {}
    _scores["Accuracy score"] = accuracy_score(Y_test, pred)
    _scores["Confusion matrix"] = conf_mat
    _scores["Precision score"] =precisionscore
    _scores["Recall score"] = recallscore
    table[name] = _scores

/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [29]:
print ("Feature Scaling: SelectKBest")
print(( "precisionscores ---- " +  str(precisionscores) +"\n" + "Accuracy score----" +str(accuracy) +"\n" +"Recall score----" + str(recallscores)+"\n"+ "Confusion matrix----"+ str(confusion_matrices)))

Feature Scaling: SelectKBest
precisionscores ---- [1.0, 0.0, 0.0, 0.5, 0.5, 1.0, 1.0, 1.0]
Accuracy score----[0.9333333333333333, 0.8666666666666667, 0.8666666666666667, 0.8666666666666667, 0.8666666666666667, 0.9333333333333333, 0.9333333333333333, 0.9333333333333333]
Recall score----[0.5, 0.0, 0.0, 0.5, 0.5, 0.5, 0.5, 0.5]
Confusion matrix----[array([[13,  0],
       [ 1,  1]]), array([[13,  0],
       [ 2,  0]]), array([[13,  0],
       [ 2,  0]]), array([[12,  1],
       [ 1,  1]]), array([[12,  1],
       [ 1,  1]]), array([[13,  0],
       [ 1,  1]]), array([[13,  0],
       [ 1,  1]]), array([[13,  0],
       [ 1,  1]])]


data in a tabular fashion

In [30]:
pd.DataFrame.from_dict(table)

K Nearest Neighbors            RBF SVM         Linear SVM  \
Accuracy score              0.933333           0.866667           0.866667   
Confusion matrix   [[13, 0], [1, 1]]  [[13, 0], [2, 0]]  [[13, 0], [2, 0]]   
Precision score                    1                  0                  0   
Recall score                     0.5                  0                  0   

                      Decision Tree        Naive Bayes           AdaBoost  \
Accuracy score             0.866667           0.866667           0.933333   
Confusion matrix  [[12, 1], [1, 1]]  [[12, 1], [1, 1]]  [[13, 0], [1, 1]]   
Precision score                 0.5                0.5                  1   
Recall score                    0.5                0.5                0.5   

                      Random Forest Logistic Regression  
Accuracy score             0.933333            0.933333  
Confusion matrix  [[13, 0], [1, 1]]   [[13, 0], [1, 1]]  
Precision score                   1                   1  
Recall score                    0.5                 0.5

ACCURACY SCORE = 0.933333

K Nearest Neighbors

AdaBoost

Random Forest

Logistic Regression

    

Taking AdaBoost as classifier for my_classifier.pkl file

In [31]:
model = AdaBoostClassifier(random_state=42)
model.fit(X_train_minmax_skb, Y_train)
print(accuracy_score(Y_test, pred))

0.9333333333333333


In [32]:
pickle.dump(data_dict, open("mydataset.pkl", "wb") )
pickle.dump(model, open("myclassifier.pkl", "wb") )
pickle.dump(modified_feature, open(" myfeaturelist.pkl", "wb") )